### PR TITLE
Intial layout and startup scripts for workbench package

### LINKED ISSUE #20381: Create initial layout for mantidqt/workbench python modules
The new [workbench design](https://github.com/mantidproject/documents/blob/master/Design/Workbench/design-contents.md) makes extensive use of a utility `mantidqt` library that will contain the reusable widgets and any required plotting code.

This issue is to create a basic layout and build rules for both the `mantidqt` package and `workbench` package, along with start up code for a main window. 

Installation will not yet be considered here.

### PR DESCRIPTION
**Pull request by martyngigg**
**Description of work**

Adds the `workbench` python package plus minimal start up code to give a functioning application. 

**To test:**

* You will need the Qt 5 libraries:
 * Windows: nothing to do
 * Linux: Update the relevant `mantid-developer` package for your platform  
 * OSX: Use `brew` to install `qt`, `sip` & `pyqt`. See instructions [here](https://github.com/mantidproject/mantid/wiki/Build-environment-setup-on-OS-X-10.12-Sierra)  
* Checkout the branch and run `cmake -DENABLE_WORKBENCH=ON .`
* Build the `workbench` target
  * it should create an egg link in the bin directory to the source code without copying it
* ctest -R workbench should run the tests
* `bin/workbench` (Linux/OSX) or `bin/CONFIG/workbench.exe` (windows) should start the application (CTRL+Q) to quit.

Fixes #20381 

**Release Notes** 

*Does not need to be in the release notes as there is nothing to shout about yet.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/python/mantidqt/py3compat.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+# -----------------------------------------------------------------------------
+# Strings
+# -----------------------------------------------------------------------------
+def is_text_string(obj):
+    """Return True if `obj` is a text string, False if it is anything else,
+    like binary data (Python 3) or QString (Python 2, PyQt API #1)"""
+    if PY2:
+        # Python 2
+        return isinstance(obj, basestring)
```
Comment (rosswhitfield) [inline qt/python/mantidqt/py3compat.py]: Why not use `six.string_types`? Then you don't need the `if`, simply `return isinstance(obj, string_types)`

File: qt/python/.gitignore [hunk 2]
```diff
@@ -0,0 +1,107 @@
+# File generated by the build
+mantidqt/utils/qt/plugins.py
```
Comment (rosswhitfield) [inline qt/python/.gitignore]: Why not just add these to the top level `.gitignore`?

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
I can't get this to start for me, I just get
```
$ bin/workbench 
Traceback (most recent call last):
  File "bin/workbench", line 11, in <module>
    load_entry_point('workbench', 'gui_scripts', 'workbench')()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 572, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2769, in load_entry_point
    return ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2422, in load
    return self.resolve()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2428, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/rwp/workbench-package-layout/qt/applications/workbench/workbench/app/mainwindow.py", line 42, in <module>
    from mantidqt.utils.qt import load_ui, plugins  # noqa
ModuleNotFoundError: No module named 'mantidqt'
```

File: qt/python/.gitignore (hunk 2 above)
Comment (martyngigg) [reply qt/python/.gitignore]: Good point. It saves confusion over where these things are.

File: qt/python/mantidqt/py3compat.py (hunk 1 above)
Comment (martyngigg) [reply qt/python/mantidqt/py3compat.py]: :+1:

**Comment by martyngigg** [comment]
It looks I'd set `PYTHONPATH` to include the bin directory and forgotten about it. It would be nice to not have to do that. Any ideas?

I could handwrite the entry scripts and force the script directory into the `sys.path`?

File: qt/python/mantidqt/utils/qt/plugins.py.in [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    """
+    QCoreApplication.addLibraryPath(
+        '@QT_PLUGINS_PATH@'.replace('%V', QT_VERSION[0])
+    )
+
+# Set the implementation appropriate for the platform
+if os.name == 'nt':
+    setup_library_paths = setup_library_paths_win
+else:
+    setup_library_paths = lambda _1, _2: None
```
Comment (rosswhitfield) [inline qt/python/mantidqt/utils/qt/plugins.py.in]: I'm getting
```
Traceback (most recent call last):
  File "bin/workbench", line 11, in <module>
    load_entry_point('workbench', 'gui_scripts', 'workbench')()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 572, in load_entry_point
    return get_distribution(dist).load_entry_point(group, name)
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2769, in load_entry_point
    return ep.load()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2422, in load
    return self.resolve()
  File "/usr/lib/python3.6/site-packages/pkg_resources/__init__.py", line 2428, in resolve
    module = __import__(self.module_name, fromlist=['__name__'], level=0)
  File "/home/rwp/workbench-package-layout/qt/applications/workbench/workbench/app/mainwindow.py", line 45, in <module>
    plugins.setup_library_paths()
TypeError: <lambda>() missing 2 required positional arguments: '_1' and '_2'
```

I changed it to `setup_library_paths = lambda: None` to get it to work.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
Setting `PYTHONPATH` I can get this to start but that's not the right solution. But I don't know what is.

**Comment by martyngigg** [comment]
@rosswhitfield I think I've figured a way around this using the `install_requires` argument to `setuptools.setup`. It seems to work for me on Linux and Windows without requiring the `PYTHONPATH`.

**Comment by martyngigg** [comment]
I'm tempted to make similar changes to the framework `mantid` package in the future and then developers may not need `mantidpython`...

File: qt/python/CMakeLists.txt [hunk 4]
```diff
@@ -5,14 +5,34 @@
...
+  configure_file ( mantidqt/utils/qt/plugins.py.in
+    ${CMAKE_CURRENT_SOURCE_DIR}/mantidqt/utils/qt/plugins.py
+  )
+
+  # Create egg link to binary output directory for mantidqt
+  add_python_package ( mantidqt )
+
+  # ctest targets
+  set ( PYTHON_TEST_FILES
+    mantidqt/test/import_test.py
```
Comment (rosswhitfield) [inline qt/python/CMakeLists.txt]: Did you mean to move this test to only run when the workbench is enabled?

Should we set `ENABLE_WORKBENCH=ON` on pull requests? We just need to install the required dependencies.

Comment (martyngigg) [reply qt/python/CMakeLists.txt]: I had originally intended to only have the `mantidqt` module in use for the workbench but following the TSC discussion there is no reason it cannot be available for both versions.

I would like to get `ENABLE_WORKBENCH=ON` for the pull requests too. Are we just waiting on PyQt5 for RHEL?

Comment (rosswhitfield) [reply qt/python/CMakeLists.txt]: 👍 We can enable it later once we get all the dependencies.

**Review by rosswhitfield — APPROVED** [review]

**Pull request merged**